src/metaxy/versioning/dict_based.py

```diff
@@ -0,0 +1,94 @@
+"""Dict-based versioning engine for databases without native struct support.
+
+This module provides a base class for versioning engines that work with
+dict[str, str] (mapping field names to column names) instead of struct columns.
+
+Instead of building struct columns like:
+    df["metaxy_provenance_by_field"] = Struct(field1="hash1", field2="hash2")
+
+We keep fields as separate columns with a naming convention:
+    df["metaxy_provenance_by_field__field1"] = "hash1"
+    df["metaxy_provenance_by_field__field2"] = "hash2"
+
+This avoids struct field access operations (struct.field()) which are not
+supported by all SQL databases in Ibis (e.g., PostgreSQL).
+
+The database store is responsible for packing/unpacking these flattened columns
+into JSON/JSONB columns at database boundaries.
+"""
+
+from __future__ import annotations
+
+from abc import ABC
+
+import narwhals as nw
+from narwhals.typing import FrameT
+
+from metaxy.models.plan import FeaturePlan
+from metaxy.versioning.engine import VersioningEngine
+from metaxy.versioning.ibis import IbisHashFn, IbisVersioningEngine
+from metaxy.versioning.types import HashAlgorithm
+
+
+class FlatVersioningEngine:
+    """Helpers for representing struct-like data as flattened columns."""
+
+    @staticmethod
+    def _get_flattened_column_name(struct_name: str, field_name: str) -> str:
+        """Get the flattened column name for a struct field."""
+        return f"{struct_name}__{field_name}"
+
+    @staticmethod
+    def build_struct_column(
```

Comment:
It's weird how a method that's called build_struct_column doesn't build any struct columns 

Should it be renamed? And I mean renamed everywhere, on the base class as well.

Reply:
it is bulding a struct in all other cases - here we are just doing something else/compatible for stores which are not able to build the struct.

It would be cleaner if we find a better super-name

Reply:
how about `record_field_versions`?